Update how Cart/Checkout is tracked to be realtime.

senadir opened the pull request:
## Proposed changes:
<!--- Explain what functional changes your PR includes -->
This PR changes how we're tracking Checkout block events. Instead of firing the event when Checkout loads, and assuming whatever is in the Checkout page is that, we now only react to Checkout events (shortcode and Block), and dynamically edit the `checkout_page_contains_checkout_block` and `checkout_page_contains_checkout_shortcode` values.

This is done by hooking into Checkout events.

This PR also adds a `from_checkout` prop to indicate if the event came from Checkout page or not.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

- Setup a new website That's running Checkout Block on the Checkout page. Install One Page Checkout, and mark a product a single page product.
- Create a new page called Checkout Shortcode and add Shortcode Checkout to it.
- Add a product to your cart, open track vigilante, and go to Checkout page (with Block), make sure the `checkout_page_contains_checkout_block` value is `1` and `checkout_page_contains_checkout_shortcode` is `0` on `product_checkout` event, make sure that `from_checkout` is `1`.
- Place the order, make sure the values are still correct in the `product_purshase` event.
- Do the same from the one page checkout product page, make sure the values are reversed, so `shortcode` is 1 and `block` is 0, and that `from_checkout` is `0`.
- Do the same from the extra page you create and make sure the values are still correct.
- Now change your default Checkout to shortcode, run the flow again the Checkout page and make sure the values are correct in both events.

senadir committed:
Update how Cart/Checkout is tracked

senadir commented on projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php:
```diff
@@ -237,30 +226,55 @@ function ( $payment_gateway ) {
+					typeof wp.data !== 'undefined' &&
+					typeof wp.data.subscribe !== 'undefined'
+				) {
+					wp.data.subscribe( function () {
```
I think I can limit it to a store here as well.

opr commented on projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php:
```diff
@@ -237,30 +226,55 @@ function ( $payment_gateway ) {
+					// This is only triggered on the checkout shortcode.
+					console.log('running here');
+					jQuery( document.body ).on( 'init_checkout', function () {
+						console.log('but not here');
```
two console logs here

opr commented on projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php:
```diff
@@ -301,6 +315,17 @@ public function order_process( $order_id ) {
+		} else {
+			$checkout_page_contains_checkout_block     = '0';
+			$checkout_page_contains_checkout_shortcode = '0';
+		}
```
I don't think it's a big deal but I would initialise these as 0 above the if in order to prevent the final `else`.

opr commented on projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php:
```diff
@@ -237,30 +226,55 @@ function ( $payment_gateway ) {
+						}
+
+						const checkoutDataStore = wp.data.select( 'wc/store/checkout' );
+						if ( undefined !== checkoutDataStore && checkoutDataStore.getOrderId() !== 0 ) {
```
I would check `if ( 'undefined' !== typeof checkoutDataStore && checkoutDataStore.getOrderId() !== 0 )` rather than using `undefined` for consistency with the above uses

senadir replied:
This is outdated I think, because I already removed the `checkoutDataStore` check because we're subscribing to it so it's always there.

senadir committed:
add extra to see from checkout is coming

senadir commented:
@anomiex let us know that the failing E2E test above is happening to all issues and it shouldn't prevent us from merging given everything else is passing.

opr commented on projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-trait.php:
```diff
@@ -107,14 +107,14 @@ function ( $payment_gateway ) {
 			'express_checkout'       => 'null', // TODO: not solved yet.
 			'products_count'         => $cart->get_cart_contents_count(),
-			'order_value'            => $cart->get_cart_total(),
+			'order_value'            => $cart_total,
```
Nice catch 🙈

opr commented on projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-checkout-flow.php:
```diff
@@ -124,10 +141,26 @@ public function capture_cart_view() {
+		global $post;
+		$checkout_page_id = wc_get_page_id( 'checkout' );
+
+		$is_checkout = $checkout_page_id && is_page( $checkout_page_id )
```
Do we also need to account for the classic shortcode block here too?

senadir replied:
Good catch, fixed.

opr reviewed (changes requested):
As a logged-in user, in `woocommerceanalytics_checkout_view` and `woocommerceanalytics_product_checkout` events the `guest_checkout` property is `Yes` for me but in `woocommerceanalytics_order_confirmation_view` it becomes `No`.

In the testing instructions:

> Do the same from the extra page you create and make sure the values are still correct.

Just noting that in this case `woocommerceanalytics_checkout_view` is showing the incorrect values. Only `woocommerceanalytics_product_checkout` and `woocommerceanalytics_product_purchase` are correct.

senadir committed:
address changes

senadir commented:
@opr I addressed the review points, added tracking to the Checkout view event as well (while trying to keep it PHP only) and made sure the event triggers regardless (previously it required you to interact with Checkout for it to run, now it should run on mount).

senadir commented on projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-checkout-flow.php:
```diff
@@ -147,18 +147,27 @@ public function capture_checkout_view() {
 		$is_checkout = $checkout_page_id && is_page( $checkout_page_id )
 		|| wc_post_content_has_shortcode( 'woocommerce_checkout' )
 		|| has_block( 'woocommerce/checkout', $post )
+		|| has_block( 'woocommerce/classic-shortcode', $post )
```
There is no way to check if the block is Cart or Checkout.

opr replied:
True, this is OK though. It's based on the template as far as I can tel. If someone has set up their Checkout page to use the Cart template then it will be an edge case, don't think we can or should try to solve for this.

opr reviewed (approved):
Happy to say this is testing OK for me, tried with several  different configs and all seem to be fine for me.

🤞🏼

jjcoffee commented on projects/plugins/jetpack/modules/woocommerce-analytics/classes/class-jetpack-woocommerce-analytics-universal.php:
```diff
@@ -237,30 +237,57 @@ function ( $payment_gateway ) {
+						if ( true === cartItem_{$cart_item_key}_logged ) {
+							return;
+						}
+						wp.hooks.addAction( 'wcpay.payment-request.availability', 'wcpay', function ( args ) {
```
`wp.hooks` is not guaranteed to be available. If it's not available, this blocks a lot of JS on the checkout from working, including coupons, "Click here to login", etc. See [here](https://wordpress.org/support/topic/woocommerce-checkout-coupon-cant-add/).

opr replied:
Thanks for raising this @jjcoffee - I agree we should add a check to ensure it's available before using.

opr replied:
@jjcoffee - I have opened https://github.com/Automattic/jetpack/pull/36560 to solve this.
